src/models/restricted_pairwise_union_reach_surface.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class for modeling Restricted Pairwise union reach surface."""
+
+import copy
+import numpy as np
+from scipy.optimize import minimize
+import cvxpy as cp
+from cvxopt import solvers
+from typing import Iterable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+from wfa_planning_evaluation_framework.models.pairwise_union_reach_surface import (
+    PairwiseUnionReachSurface,
+)
+
+
+class RestrictedPairwiseUnionReachSurface(PairwiseUnionReachSurface):
+    """Models reach with the pairwise union overlap model."""
+
+    def __init__(
+        self, reach_curves: Iterable[ReachCurve], reach_points: Iterable[ReachPoint]
+    ):
+        """Constructor for RestrictedPaiwiseUnionReachSurface.
+
+        Args:
+          reach_curves: A list of ReachCurves to be used in model fitting and reach
+            prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit. This
+            is parallel to the reach_curves list. The reach point at ith poisition
+            is drawn from ith reach curve.
+        """
+        super().__init__(reach_curves=reach_curves, reach_points=reach_points)
+
+    def _fit(self) -> None:
+        self._reach_vectors = np.array(
+            [
+                self.get_reach_vector(reach_point.impressions)
+                for reach_point in self._data
+            ]
+        )
+        cons = self.get_constraints()
+        res = minimize(
+            fun=lambda x: self.loss(x),
+            x0=np.array([2 / self._p] * self._p) / 2,
+            constraints=cons,
+            method="trust-constr",
+            options={"disp": True},
+        )
+        self.construct_a_from_lambda(res["x"])
+
+    def construct_a_from_lambda(self, lbd):
+        """Get value of flattened a matrix from lamdas.
+
+        Args:
+          lbd: a length p vector indicating lambda_i for each pub.
+
+        Returns:
+          the value of flattened a matrix.
+        """
+        self._a = np.ones(self._p * self._p)
```

Comment:
I suppose this function could be written as:

p = len(lbd)
A = lbd.reshape(p,1) * lbd.reshape(1,p) - (np.eye(p) * lbd)**2
return A.reshape(,1)

Reply:
thanks